aggregator/src/aggregator/collection_job_driver.rs

```diff
@@ -200,9 +214,50 @@ impl CollectionJobDriver {
                 leader_aggregate_share,
             }),
         );
+        let batch_aggregations = Arc::new(batch_aggregations);
+
+        // To ensure that concurrent aggregations don't write into a currently-nonexistent batch
+        // aggregation, we write (empty) batch aggregations for any that have not already been
+        // written to storage.
+        let existing_batch_aggregations: HashSet<_> = batch_aggregations
+            .iter()
+            .map(|ba| (ba.batch_identifier(), ba.ord()))
+            .collect();
+        let empty_batch_aggregations: Arc<Vec<_>> = Arc::new(
+            iproduct!(
+                Q::batch_identifiers_for_collect_identifier(
+                    &task,
+                    collection_job.batch_identifier()
+                ),
+                0..self.batch_aggregation_shard_count
+            )
+            .filter_map(|(batch_identifier, ord)| {
+                if !existing_batch_aggregations.contains(&(&batch_identifier, ord)) {
+                    Some(BatchAggregation::<SEED_SIZE, Q, A>::new(
+                        *task.id(),
+                        batch_identifier,
+                        collection_job.aggregation_parameter().clone(),
+                        ord,
+                        BatchAggregationState::Collected,
+                        None,
+                        0,
+                        Interval::EMPTY,
+                        ReportIdChecksum::default(),
+                    ))
+                } else {
+                    None
+                }
+            })
+            .collect(),
```

Comment:
nit: this does basically the exact same thing as a similar codepath on the Helper side in `aggregator.rs` and so maybe could be factored out. I'm not sure what you'd hang the new method on, though.